coord/src/rmr/command.c

```diff
@@ -342,41 +171,13 @@ void MRCommand_ReplaceArg(MRCommand *cmd, int index, const char *newArg, size_t
   MRCommand_ReplaceArgNoDup(cmd, index, news, len);
 }
 
-MRCommandFlags MRCommand_GetFlags(MRCommand *cmd) {
-  if (cmd->id < 0) return 0;
-  return __commandConfig[cmd->id].flags;
-}
-
-MRCommandGenerator *MRCommand_GetCommandGenerator(MRCommand *cmd) {
-  if (cmd->id < 0) {
-    return NULL;  // default
-  }
-
-  return __commandConfig[cmd->id].commandGenerator;
-}
-
-int MRCommand_GetShardingKey(MRCommand *cmd) {
-  if (cmd->id < 0) {
-    return 1;  // default
-  }
-
-  return __commandConfig[cmd->id].keyPos;
-}
-
-int MRCommand_GetPartitioningKey(MRCommand *cmd) {
-  if (cmd->id < 0) {
-    return 1;  // default
-  }
-
-  return __commandConfig[cmd->id].partitionKeyPos;
-}
-
-/* Return 1 if the command should not be sharded (i.e a coordination command or system command) */
-int MRCommand_IsUnsharded(MRCommand *cmd) {
-  if (cmd->id < 0) {
-    return 0;  // default
-  }
-  return __commandConfig[cmd->id].keyPos <= 0;
+// Should only be relevant for _FT.ADD, _FT.GET, _FT.DEL,
+// and _FT.SUG* commands
+int MRCommand_GetShardingKey(const MRCommand *cmd) {
+  // for SUGADD, SUGGET, SUGDEL, SUGLEN, the key is the first argument
+  // for ADD, GET, DEL, the key is the second argument.
+  // Differentiate between the two cases by checking the command length
+  return (cmd->lens[0] == 7) ? 2 : 1;
```

Comment:
I would honestly prefer to create a simple map maybe? I think that checking command len is fragile and feels wrong.

Reply:
In general, I agree. The idea of RediSearch's OSS-cluster supporting redirection of key/slot bounded commands is bad and we shouldn't do it, but I don't want to break it now.
I can't imagine a new command getting here, so although it is fragile, for performance reasons I think it is better

Reply:
No not dead yet.
RediSearch has 3 types of commands:
1. commands that should be fan-out to all shards
2. commands that can be answered by every shard
3. commands that are designated to a specific shard

Most of the commands are of the first two types, and after this PR this is also the way we answer them (previously we redirected some of the commands of the second type, sometimes to some shard, sometimes to the shard in the first partition).

the only commands that are of the third type are the `FT.SUG*` family, which are all key-bounded commands, and the `FT.ADD`, `FT.GET`, and `FT.DEL` that should be deprecated, but also key-bounded. On redis OSS these commands should return a `MOVED` error when they are sent to the wrong shard, but RediSearch implements the workaround and redirects them to the right shard. They are the only reason the coordinator cares about the sharding function, so it could calculate the right destination for these commands.

No new commands should get this treatment IMO, so we can use this optimized function to handle only them. I agree that having here a long `if else if...` list of `strcasecmp` is more human-readable, but it will be costly and unnecessary.